pikelet/src/surface/lex.rs

```diff
@@ -0,0 +1,190 @@
+            T::Number(s) => write!(f, "Number({})", s),
+            T::Name(s) => write!(f, "Write({})", s),
+            T::Shift(s) => write!(f, "Shift({})", s),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct LexicalError(std::ops::Range<usize>, &'static str);
+
+impl Display for LexicalError {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "Lexical error: {:?} {}", self.0, self.1)
+    }
+}
+
+pub struct LexIterator<'a>(logos::Lexer<LexToken, &'a str>);
+pub type Spanned<Tok, Loc, Error> = Result<(Loc, Tok, Loc), Error>;
+
+impl<'a> LexIterator<'a> {
+    pub fn new(source: &'a str) -> LexIterator<'a> {
+        LexIterator(LexToken::lexer(source))
+    }
+}
+
+impl<'a> Iterator for LexIterator<'a> {
+    type Item = Spanned<Token<'a>, usize, LexicalError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        use LexToken as LT;
+        use Token as T;
+        let lex = &mut self.0;
+
+        while let LT::Whitespace = &lex.token {
+            lex.advance();
+        }
+
+        const fn tok<'a>(
+            r: std::ops::Range<usize>,
+            t: Token<'a>,
+        ) -> Option<Spanned<Token<'a>, usize, LexicalError>> {
+            Some(Ok((r.start, t, r.end)))
+        }
+
+        let r = lex.range();
+        // I'm really not fond of this, but trying to split this match statement up
+        // appears to be more trouble than it is worth.
+        #[rustfmt::skip]
+        let it: Option<Self::Item> = match &lex.token {
+            LT::EOF => None,
+            LT::Error => Some(Err(LexicalError(r, "Lexical error"))),
+            LT::Whitespace => unreachable!(),
+            LT::Colon  => tok(r, T::Colon),
+            LT::Comma  => tok(r, T::Comma),
+            LT::Fun    => tok(r, T::Fun),
+            LT::DArrow => tok(r, T::DArrow),
+            LT::Arrow  => tok(r, T::Arrow),
+            LT::LParen => tok(r, T::LParen),
+            LT::RParen => tok(r, T::RParen),
+            LT::LBrack => tok(r, T::LBrack),
+            LT::RBrack => tok(r, T::RBrack),
+            LT::LBrace => tok(r, T::LBrace),
+            LT::RBrace => tok(r, T::RBrace),
+            LT::Dot    => tok(r, T::Dot),
+            LT::Equal  => tok(r, T::Equal),
+            LT::RecordTerm => tok(r, T::RecordTerm),
+            LT::RecordType => tok(r, T::RecordType),
+            LT::Name    => tok(r, T::Name(lex.slice())),
+            // We could do something here besides expose this as a string if desired,
+            // that doesn't work with the way that Literal is though...
+            LT::Number  => tok(r, T::Number(lex.slice())),
+            LT::Shift   => tok(r, T::Shift(lex.slice())),
+            // I think as-is the next 2 will include the pair of quotation marks.
+            // These probably need to be stripped from CharLit and StrLit?
+            LT::CharLit => tok(r, T::CharLit(lex.slice())),
+            LT::StrLit  => tok(r, T::StrLit(lex.slice())),
```

Comment:
Ultimately I want to do the final lexing of these in the elaborator (ie. `surface::projections::core`), with the benefit of type information. So this should be fine I think?

Reply:
Ok, I just removed those comments then.